packages/flutter/lib/src/widgets/widget_inspector.dart

```diff
@@ -0,0 +1,444 @@
+
+  final Size tooltipSize =
+      textPainter.size + const Offset(_kTooltipPadding * 2, _kTooltipPadding * 2);
+  final _TooltipPositionCalculator calc = new _TooltipPositionCalculator(
+      target: target,
+      verticalOffset: verticalOffset,
+      preferBelow: false,
+      size: size,
+      childSize: tooltipSize);
+  final Offset tipOffset = calc.childPosition;
+
+  final Paint tooltipBackground = new Paint()
+    ..style = PaintingStyle.fill
+    ..color = _kTooltipBackgroundColor;
+  canvas.drawRect(
+      new Rect.fromPoints(tipOffset,
+          tipOffset.translate(tooltipSize.width, tooltipSize.height)),
+      tooltipBackground);
+
+  double wedgeY = tipOffset.dy;
+  if (!calc.tooltipBelow) {
+    wedgeY += tooltipSize.height;
+  }
+  final double wedgeSize = _kTooltipPadding * 2;
+  double wedgeX = math.max(tipOffset.dx, target.dx) + wedgeSize * 2;
+  wedgeX = math.min(wedgeX, tipOffset.dx + tooltipSize.width - wedgeSize * 2);
+  canvas.drawPath(
+      new Path()
+        ..addPolygon(<Offset>[
+          new Offset(wedgeX - wedgeSize, wedgeY),
+          new Offset(wedgeX + wedgeSize, wedgeY),
+          new Offset(
+              wedgeX, wedgeY + (calc.tooltipBelow ? -wedgeSize : wedgeSize)),
+        ], true),
+      tooltipBackground);
+
+  textPainter.paint(
+      canvas, tipOffset + const Offset(_kTooltipPadding, _kTooltipPadding));
+  canvas.restore();
+}
+
+final Color _kHighlightedRenderObjectFillColor =
+    const Color.fromARGB(128, 128, 128, 255);
+final Color _kHighlightedRenderObjectBorderColor =
+    const Color.fromARGB(128, 64, 64, 128);
+
+class _WidgetInspectorPainter extends CustomPainter {
+  const _WidgetInspectorPainter(this.owner, this.generation, this.selected);
+
+  final PipelineOwner owner;
+  final int generation;
+  final RenderBox selected;
+
+  bool get isSelectionActive => selected != null && selected.attached;
+
+  @override
+  void paint(Canvas canvas, Size size) {
+    if (!isSelectionActive)
+      return;
+
+    canvas.save();
+
+    final Matrix4 transform = selected.getTransformTo(null);
+    // Highlight the selected renderObject.
+    canvas
+      ..save()
+      ..transform(transform.storage)
+      ..drawRect(
+          (Offset.zero & selected.size).deflate(0.5),
+          new Paint()
+            ..style = PaintingStyle.fill
+            ..color = _kHighlightedRenderObjectFillColor)
+      ..drawRect(
+          (Offset.zero & selected.size).deflate(0.5),
+          new Paint()
+            ..style = PaintingStyle.stroke
+            ..strokeWidth = 1.0
+            ..color = _kHighlightedRenderObjectBorderColor)
+      ..restore();
+
+    final List<Vector3> corners = <Vector3>[
+      transform.transform3(new Vector3(0.0, 0.0, 0.0)),
+      transform.transform3(new Vector3(selected.size.width, 0.0, 0.0)),
+      transform.transform3(
+          new Vector3(selected.size.width, selected.size.height, 0.0)),
```

Comment:
i don't think the newline helps readability here, it just makes the code less uniform